pkg/client/openstack/kluster/client.go

```diff
@@ -190,94 +202,249 @@ func (c *klusterClient) SetSecurityGroup(sgName, nodeID string) (err error) {
+		return nil, fmt.Errorf("Loadbalancer SecurityGroup %s listing error: %w", lbSgName, err)
 	}
 
-	if kluster.ClusterCIDR() == "" {
-		return false, errors.New("Cluster CIDR for kluster not set")
+	lbGroups, err := securitygroups.ExtractGroups(lbSgPage)
+	if err == nil && len(lbGroups) == 1 {
+		return &lbGroups[0], nil
 	}
 
-	groups, err := securitygroups.ExtractGroups(page)
+	return nil, err
+}
+
+func createLoadbalancerSecurityGroup(kluster *v1.Kluster, client *gophercloud.ServiceClient) (*securitygroups.SecGroup, error) {
+	sg, err := securitygroups.Create(client, securitygroups.CreateOpts{
+		Name:        fmt.Sprintf(LoadBalancerSecurityGroupName, kluster.Spec.Name),
+		Description: fmt.Sprintf(`Kubernikus: SecurityGroup containing LoadBalancer Self-IPs for cluster "%s"`, kluster.Spec.Name),
+	}).Extract()
 	if err != nil {
-		return false, err
+		return nil, err
 	}
 
-	if len(groups) != 1 {
-		return false, fmt.Errorf("More than one SecurityGroup with name %v found", sgName)
+	return sg, nil
+}
+
+func (c *klusterClient) EnsureLoadbalancerSecurityGroup(kluster *v1.Kluster, recorder record.EventRecorder) (created bool, err error) {
+	var lbGroup *securitygroups.SecGroup
+	lbGroup, err = getLoadbalancerSecurityGroup(kluster, c.NetworkClient)
+	if err != nil {
+		return false, err
 	}
 
-	udp := false
-	tcp := false
-	icmp := false
-	for _, rule := range groups[0].Rules {
-		if rule.Direction != string(rules.DirIngress) {
-			continue
+	if lbGroup == nil {
+		lbGroup, err = createLoadbalancerSecurityGroup(kluster, c.NetworkClient)
+		if err != nil || lbGroup == nil {
+			recorder.Eventf(kluster, core_v1.EventTypeWarning, events.FailedCreateLoadBalancerSecurityGroup, "Cannot create security group for loadbalancer Self-IPs. Please check for sufficient quota.")
+			return false, fmt.Errorf("Could not create loadbalancer security group: %w", err)
 		}
+	}
 
-		if rule.EtherType != string(rules.EtherType4) {
-			continue
-		}
+	allLbPages, err := loadbalancers.List(c.NetworkClient, loadbalancers.ListOpts{}).AllPages()
+	if err != nil {
+		return false, err
+	}
 
-		if rule.RemoteIPPrefix != *kluster.Spec.ClusterCIDR {
-			continue
-		}
+	allLoadbalancers, err := loadbalancers.ExtractLoadBalancers(allLbPages)
+	if err != nil {
+		return false, err
+	}
 
-		if rule.Protocol == string(rules.ProtocolICMP) {
-			icmp = true
-			continue
-		}
+	allPortPages, err := ports.List(c.NetworkClient, portListOpts{ProjectID: kluster.Labels["account"], SecurityGroups: lbGroup.ID}).AllPages()
+	if err != nil {
+		return false, err
+	}
 
-		if rule.Protocol == string(rules.ProtocolUDP) {
-			udp = true
-			continue
-		}
+	allPorts, err := ports.ExtractPorts(allPortPages)
+	if err != nil {
+		return false, err
+	}
 
-		if rule.Protocol == string(rules.ProtocolTCP) {
-			tcp = true
-			continue
+	sg := []string{lbGroup.ID}
+	for _, lb := range allLoadbalancers {
+		if strings.HasPrefix(lb.Name, fmt.Sprintf("kube_service_%s", kluster.Spec.Name)) {
+			found := false
```

Comment:
I suggest to replace the inner loop with

```
if !portsOfSecurityGroup.Has(lb.VipPortID) {
  //update port 
}
```
And to this before the whole loop construct:
```
portsOfSecurityGroup := sets.NewString()
for _, p := range allPorts {
  portsOfSecurityGroup.Insert(p.ID)
}
```
using https://pkg.go.dev/k8s.io/apimachinery/pkg/util/sets